lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,262 @@
+		private static $instance = null;
+
+		/**
+		 * Registers a block template.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string|WP_Block_Template $template_name Block template name including namespace, or alternatively
+		 *                                                a complete WP_Block_Template instance. In case a WP_Block_Template
+		 *                                                is provided, the $args parameter will be ignored.
+		 * @param array                    $args          Optional. Array of block template arguments.
+		 * @return WP_Block_Template|false The registered block template on success, or false on failure.
+		 */
+		public function register( $template_name, $args = array() ) {
+
+			$template = null;
+			if ( $template_name instanceof WP_Block_Template ) {
+				$template      = $template_name;
+				$template_name = $template->name;
+			}
+
+			if ( ! is_string( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must be strings.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( preg_match( '/[A-Z]+/', $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must not contain uppercase characters.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			$name_matcher = '/^[a-z0-9-]+\/\/[a-z0-9-]+$/';
+			if ( ! preg_match( $name_matcher, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( $this->is_registered( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					/* translators: %s: Template name. */
+					sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ),
+					'6.7.0'
+				);
+				return false;
+			}
+
+			if ( ! $template ) {
+				$theme_name = get_stylesheet();
+				$slug       = isset( $args['slug'] ) ? $args['slug'] : explode( '//', $template_name )[1];
+
+				$template                 = new WP_Block_Template();
+				$template->id             = $theme_name . '//' . $slug;
+				$template->theme          = $theme_name;
+				$template->plugin         = isset( $args['plugin'] ) ? $args['plugin'] : '';
+				$template->author         = null;
+				$template->content        = isset( $args['content'] ) ? $args['content'] : '';
+				$template->source         = 'plugin';
+				$template->slug           = $slug;
+				$template->type           = 'wp_template';
+				$template->title          = isset( $args['title'] ) ? $args['title'] : '';
+				$template->description    = isset( $args['description'] ) ? $args['description'] : '';
+				$template->status         = 'publish';
+				$template->has_theme_file = true;
+				$template->origin         = 'plugin';
+				$template->is_custom      = true;
+				$template->post_types     = isset( $args['post_types'] ) ? $args['post_types'] : '';
```

Comment:
It would be good to have something similar to [`register_taxonomy_for_object_type()`](https://developer.wordpress.org/reference/functions/register_taxonomy_for_object_type/) to allow devs to extend this list after registration.

Reply:
That sounds like a good idea. Do you think it should be part of this PR? I'm a bit wary of increasing its scope, and I think that's something that can be worked on afterwards. What do you think?

Reply:
Yeah, probably better to explore in a different PR.